File: Уник/5ый сем/АиСД/SplayTree/cpp/SplayTree.cpp
#include <iostream>
#include <string>
#include <sstream>
#include <vector>

struct Node {
    long long key;
    std::string value;
    Node* left = nullptr;
    Node* right = nullptr;
    Node* father = nullptr;
};

struct Tree {
    Node* root = nullptr;

    void rightRotate(Node* y) {
        Node* x = y->left;
        if (x == nullptr) {
            return;
        }
        y->left = x->right;
        if (x->right) {
            x->right->father = y;
        }
        x->father = y->father;
        if (y->father == nullptr) {
            root = x;
        } else if (y->father->left == y) {
            y->father->left = x;
        } else {
            y->father->right = x;
        }
        x->right = y;
        y->father = x;
    }

    void leftRotate(Node* x) {
        Node* y = x->right;
        if (y == nullptr) {
            return;
        }
        x->right = y->left;
        if (y->left) {
            y->left->father = x;
        }
        y->father = x->father;
        if (x->father == nullptr) {
            root = y;
        } else if (x->father->left == x) {
            x->father->left = y;
        } else {
            x->father->right = y;
        }
        y->left = x;
        x->father = y;
    }

    void splay(Node* x) {
        while (x->father) {
            Node* p = x->father;
            Node* g = p->father;
            if (g == nullptr) { // zig
                if (x == p->left) {
                    rightRotate(p);
                } else  {
                    leftRotate(p);
                }
            } else if (x == p->left && p == g->left) { // zig-zig левый
                rightRotate(g);
                rightRotate(p);
            } else if (x == p->right && p == g->right) { // zig-zig правый
                leftRotate(g);
                leftRotate(p);
            } else if (x == p->right && p == g->left) { // zig-zag 2 штуки
                leftRotate(p);
                rightRotate(g);
            } else {
                rightRotate(p);
                leftRotate(g);
            }
        }
    }

    Node* search(long long key) {
        if (root == nullptr) {
            return nullptr;
        }
        Node* cur = root;
        Node* lastNotNull = nullptr;
        while (cur) {
            lastNotNull = cur;
            if (key == cur->key) {
                splay(cur);
                return cur;
            } else if (key < cur->key) {
                cur = cur->left;
            } else {
                cur = cur->right;
            }
        }
        if (lastNotNull) {
            splay(lastNotNull);
        }
        return nullptr;
    }

    void add(long long key, const std::string& value) {
        if (root == nullptr) {
            root = new Node{key, value};
            return;
        }
        Node* cur = root;
        Node* lastNotNull = nullptr;
        while (cur) {
            lastNotNull = cur;
            if (key == cur->key) {
                std::cout << "error" << std::endl;
                splay(cur);
                return;
            }
            if (key < cur->key) {
                cur = cur->left;
            } else {
                cur = cur->right;
            }
        }
        Node* newNode = new Node{key, value, nullptr, nullptr, lastNotNull};
        if (key < lastNotNull->key)
            lastNotNull->left = newNode;
        else
            lastNotNull->right = newNode;
        splay(newNode);
    }

    void set(long long key, const std::string& value) {
        Node* node = search(key);
        if (node == nullptr) {
            std::cout << "error" << std::endl;
            return;
        }
        node->value = value;
    }

    void del(long long key) {
        Node* node = search(key);
        if (node == nullptr) {
            std::cout << "error" << std::endl;
            return;
        }

        Node* leftTree = node->left;
        Node* rightTree = node->right;

        if (leftTree) {
            leftTree->father = nullptr;
        }
        if (rightTree) {
            rightTree->father = nullptr;
        }

        delete node;

        if (leftTree == nullptr) {
            root = rightTree;
        } else {
            Node* maxLeft = leftTree;
            while (maxLeft->right) maxLeft = maxLeft->right;
            splay(maxLeft);
            maxLeft->right = rightTree;
            if (rightTree) rightTree->father = maxLeft;
            root = maxLeft;
        }
    }

    Node* min() {
        if (root == nullptr) {
            return nullptr;
        }
        Node* cur = root;
        while (cur->left) {
            cur = cur->left;
        }
        splay(cur);
        return cur;
    }

    Node* max() {
        if (root == nullptr) {
            return nullptr;
        }
        Node* cur = root;
        while (cur->right) {
            cur = cur->right;
        }
        splay(cur);
        return cur;
    }

    void print() {
        if (root == nullptr) {
            std::cout << "_" << std::endl;
            return;
        }

        std::vector<Node*> level = {root};
        std::cout << "[" << root->key << " " << root->value << "]" << std::endl;

        while (true) {
            std::vector<Node*> next;
            bool any = false;
            for (Node* n : level) {
                if (n) {
                    next.push_back(n->left);
                    next.push_back(n->right);
                    if (n->left || n->right) any = true;
                } else {
                    next.push_back(nullptr);
                    next.push_back(nullptr);
                }
            }
            if (!any) return;
            bool first = true;
            for (Node* n : next) {
                if (!first) std::cout << " ";
                first = false;
                if (n == nullptr) {
                    std::cout << "_";
                } else {
                    std::cout << "[" << n->key << " " << n->value << " " << n->father->key << "]";
                }
            }
            std::cout << std::endl;
            level.swap(next);
        }
    }
};

int main() {
    Tree tree;
    std::string line;

    while (std::getline(std::cin, line)) {
        if (line.empty()) continue;
        if (line.front() == ' ' || line.back() == ' ' || line.find("  ") != std::string::npos) {
            std::cout << "error" << std::endl;
            continue;
        }

        std::stringstream ss(line);
        std::string cmd;
        ss >> cmd;

        if (cmd == "add") {
            long long key;
            std::string val;
            if (!(ss >> key >> val)) {
                std::cout << "error" << std::endl;
                continue;
            }
            tree.add(key, val);
        }
        else if (cmd == "set") {
            long long key;
            std::string val;
            if (!(ss >> key >> val)) {
                std::cout << "error" << std::endl;
                continue;
            }
            tree.set(key, val);
        }
        else if (cmd == "delete") {
            long long key;
            if (!(ss >> key)) {
                std::cout << "error" << std::endl;
                continue;
            }
            tree.del(key);
        }
        else if (cmd == "search") {
            long long key;
            if (!(ss >> key)) {
                std::cout << "error" << std::endl;
                continue;
            }
            Node* res = tree.search(key);
            if (!res) std::cout << "0" << std::endl;
            else std::cout << "1 " << res->value << std::endl;
        }
        else if (cmd == "min") {
            Node* res = tree.min();
            if (!res) std::cout << "error" << std::endl;
            else std::cout << res->key << " " << res->value << std::endl;
        }
        else if (cmd == "max") {
            Node* res = tree.max();
            if (!res) std::cout << "error" << std::endl;
            else std::cout << res->key << " " << res->value << std::endl;
        }
        else if (cmd == "print") {
            tree.print();
        }
        else {
            std::cout << "error" << std::endl;
        }
    }

    return 0;
}

File: Уник/5ый сем/АиСД/SplayTree/SplayTree.go
package main

import (
	"bufio"
	"fmt"
	"os"

	"runtime"
	"strconv"
	"strings"
	// "time"
)

type Node struct {
	key    int
	value  string
	left   *Node
	right  *Node
	father *Node
}

type Tree struct {
	root *Node
}

func (tree *Tree) zig(x *Node) {
	p := x.father
	if p.father == nil {
		tree.root = x
		x.father = nil
	} else {
		g := p.father
		if p == g.left {
			g.left = x
		} else {
			g.right = x
		}
		x.father = g
	}
	if x == p.left {
		p.left = x.right
		if x.right != nil {
			x.right.father = p
		}
		x.right = p
	} else {
		p.right = x.left
		if x.left != nil {
			x.left.father = p
		}
		x.left = p
	}
	p.father = x
}

func (tree *Tree) zig_zig(x *Node) {
	tree.zig(x.father)
	tree.zig(x)
}

func (tree *Tree) zig_zag(x *Node) {
	tree.zig(x)
	tree.zig(x)
}

func (tree *Tree) splay(x *Node) {
	for x.father != nil {
		p := x.father
		if p.father == nil {
			tree.zig(x)
		} else if (x == p.left) == (p == p.father.left) {
			tree.zig_zig(x)
		} else {
			tree.zig_zag(x)
		}
	}
}

func (tree *Tree) search(key int) *Node {
	if tree.root == nil {
		return nil
	}
	cur := tree.root
	var lastNotNil *Node
	for cur != nil {
		lastNotNil = cur
		if key == cur.key {
			tree.splay(cur)
			return cur
		}
		if key < cur.key {
			cur = cur.left
		} else {
			cur = cur.right
		}
	}
	if lastNotNil != nil {
		tree.splay(lastNotNil)
	}
	return nil
}

func (tree *Tree) add(key int, value string) {
	if tree.root == nil {
		tree.root = &Node{key: key, value: value}
		return
	}
	cur := tree.root
	var lastNotNil *Node
	for cur != nil {
		lastNotNil = cur
		if key == cur.key {
			fmt.Println("error")
			tree.splay(cur)
			return
		}
		if key < cur.key {
			cur = cur.left
		} else {
			cur = cur.right
		}
	}
	newNode := &Node{key: key, value: value, father: lastNotNil}
	if key < lastNotNil.key {
		lastNotNil.left = newNode
	} else {
		lastNotNil.right = newNode
	}
	tree.splay(newNode)
}

func (tree *Tree) set(key int, value string) {
	node := tree.search(key)
	if node == nil {
		fmt.Println("error")
		return
	}
	node.value = value
}

func (tree *Tree) delete(key int) {
	node := tree.search(key)
	if node == nil {
		fmt.Println("error")
		return
	}

	leftSub := node.left
	rightSub := node.right

	if leftSub != nil {
		leftSub.father = nil
	}
	if rightSub != nil {
		rightSub.father = nil
	}

	if leftSub == nil {
		tree.root = rightSub
	} else {
		maxLeft := leftSub
		for maxLeft.right != nil {
			maxLeft = maxLeft.right
		}
		tree.splay(maxLeft)
		maxLeft.right = rightSub
		if rightSub != nil {
			rightSub.father = maxLeft
		}
		tree.root = maxLeft
	}
}

func (tree *Tree) min() *Node {
	if tree.root == nil {
		return nil
	}
	cur := tree.root
	for cur.left != nil {
		cur = cur.left
	}
	tree.splay(cur)
	return cur
}

func (tree *Tree) max() *Node {
	if tree.root == nil {
		return nil
	}
	cur := tree.root
	for cur.right != nil {
		cur = cur.right
	}
	tree.splay(cur)
	return cur
}

func (tree *Tree) print() {
	if tree.root == nil {
		fmt.Println("_")
		return
	}
	var sb strings.Builder
	fmt.Fprintf(&sb, "[%d %s]\n", tree.root.key, tree.root.value)

	queue := []any{}
	if tree.root.left != nil {
		queue = append(queue, tree.root.left)
	} else {
		queue = append(queue, 1)
	}

	if tree.root.right != nil {
		queue = append(queue, tree.root.right)
	} else {
		switch queue[len(queue)-1].(type) {
		case int:
			return
		case *Node:
			queue = append(queue, 1)
		}
	}

	anyOne := true
	for anyOne {
		anyOne = false
		next := []any{}
		for i := 0; i < len(queue)-1; i++ {
			switch v := queue[i].(type) {
			case int:
				fmt.Fprint(&sb, strings.Repeat("_ ", v))
				if len(next) == 0 {
					next = append(next, 2*v)
				} else {
					switch z := next[len(next)-1].(type) {
					case *Node:
						if len(next) != 0 {
							next = append(next, 2*v)
						}
					case int:
						if len(next) != 0 {
							next[len(next)-1] = z + 2*v
						}
					}
				}
			case *Node:
				fmt.Fprintf(&sb, "[%d %s %d] ", v.key, v.value, v.father.key)
				if v.left != nil {
					next = append(next, v.left)
					anyOne = true
				} else {
					if len(next) == 0 {
						next = append(next, 1)
					} else {
						switch z := next[len(next)-1].(type) {
						case int:
							next[len(next)-1] = z + 1
						case *Node:
							if len(next) != 0 {
								next = append(next, 1)
							}
						}
					}
				}
				if v.right != nil {
					next = append(next, v.right)
					anyOne = true
				} else {
					if len(next) == 0 {
						next = append(next, 1)
					} else {
						switch z := next[len(next)-1].(type) {
						case int:
							next[len(next)-1] = z + 1
						case *Node:
							if len(next) != 0 {
								next = append(next, 1)
							}
						}
					}
				}
			}
		}
		// вывод последнего
		temp := queue[len(queue)-1]
		switch v := temp.(type) {
		case int:
			fmt.Fprint(&sb, strings.Repeat("_ ", v-1))
			fmt.Fprint(&sb, "_\n")
			if len(next) == 0 {
				next = append(next, 2*v)
			} else {
				switch z := next[len(next)-1].(type) {
				case *Node:
					if len(next) != 0 {
						next = append(next, 2*v)
					}
				case int:
					if len(next) != 0 {
						next[len(next)-1] = z + 2*v
					}
				}
			}
		case *Node:
			fmt.Fprintf(&sb, "[%d %s %d]\n", v.key, v.value, v.father.key)
			if v.left != nil {
				next = append(next, v.left)
				anyOne = true
			} else {
				if len(next) == 0 {
					next = append(next, 1)
				} else {
					switch z := next[len(next)-1].(type) {
					case int:
						next[len(next)-1] = z + 1
					case *Node:
						if len(next) != 0 {
							next = append(next, 1)
						}
					}
				}
			}
			if v.right != nil {
				next = append(next, v.right)
				anyOne = true
			} else {
				if len(next) == 0 {
					next = append(next, 1)
				} else {
					switch z := next[len(next)-1].(type) {
					case int:
						next[len(next)-1] = z + 1
					case *Node:
						if len(next) != 0 {
							next = append(next, 1)
						}
					}
				}
			}
		}
		queue = next
	}
	fmt.Print(sb.String())
}

func main() {
	tree := &Tree{}
	scanner := bufio.NewScanner(os.Stdin)

	for scanner.Scan() {
		line := scanner.Text()
		if line == "" {
			continue
		}
		fields := strings.Fields(line)
		switch fields[0] {
		case "add":
			if len(fields) == 3 {
				key, err := strconv.Atoi(fields[1])
				if err != nil {
					fmt.Println("error")
					continue
				}
				tree.add(key, fields[2])
			} else if len(fields) == 2 {
				key, err := strconv.Atoi(fields[1])
				if err != nil {
					fmt.Println("error")
					continue
				}
				tree.add(key, "")
			} else {
				fmt.Println("error")
				continue
			}

		case "set":
			if len(fields) == 3 {
				key, err := strconv.Atoi(fields[1])
				if err != nil {
					fmt.Println("error")
					continue
				}
				tree.set(key, fields[2])
			} else if len(fields) == 2 {
				key, err := strconv.Atoi(fields[1])
				if err != nil {
					fmt.Println("error")
					continue
				}
				tree.set(key, "")
			} else {
				fmt.Println("error")
				continue
			}

		case "delete":
			if len(fields) != 2 {
				fmt.Println("error")
				continue
			}
			key, err := strconv.Atoi(fields[1])
			if err != nil {
				fmt.Println("error")
				continue
			}
			tree.delete(key)

		case "search":
			if len(fields) != 2 {
				fmt.Println("error")
				continue
			}
			key, err := strconv.Atoi(fields[1])
			if err != nil {
				fmt.Println("error")
				continue
			}
			temp := tree.search(key)
			if temp == nil {
				fmt.Println("0")
			} else {
				fmt.Printf("1 %s\n", temp.value)
			}

		case "max":
			if len(fields) != 1 {
				fmt.Println("error")
				continue
			}

			temp := tree.max()
			if temp == nil {
				fmt.Println("error")
				continue
			}
			fmt.Printf("%d %s\n", temp.key, temp.value)

		case "min":
			if len(fields) != 1 {
				fmt.Println("error")
				continue
			}

			temp := tree.min()
			if temp == nil {
				fmt.Println("error")
				continue
			}
			fmt.Printf("%d %s\n", temp.key, temp.value)

		case "print":
			tree.print()
		default:
			fmt.Println("error")
		}
	}
	var m runtime.MemStats
	runtime.ReadMemStats(&m)
	fmt.Printf("Alloc = %v KB\n", m.Alloc/1024)
	fmt.Printf("TotalAlloc = %v KB\n", m.TotalAlloc/1024)
	fmt.Printf("Sys = %v KB\n", m.Sys/1024)
	fmt.Printf("NumGC = %v\n", m.NumGC)
}
